src/rred_reports/reports/auth.py

```diff
@@ -1,36 +1,86 @@
 """Mail server authentication"""
+import logging
 from dataclasses import dataclass
+from pathlib import Path
+from urllib import parse
 
+import click
+import msal
 from dynaconf.base import Settings
 from exchangelib import (
     DELEGATE,
     Account,
     Configuration,
-    OAuth2LegacyCredentials,
+    OAuth2AuthorizationCodeCredentials,
 )
+from oauthlib.oauth2 import OAuth2Token
 
 from rred_reports.reports import get_settings
 
+LOG_FORMAT = "%(levelname)-10s %(asctime)s %(name)-30s %(funcName)-35s %(lineno)-5d: %(message)s"
+logging.basicConfig(level=logging.INFO, format=LOG_FORMAT)
+logger = logging.getLogger(__name__)
+
+CACHE_PATH = Path("my_cache.bin")
+
 
 @dataclass
 class RREDAuthenticator:
-    """Class to handle MS exchange server authentication"""
+    """Class to handle MS Exchange server authentication"""
 
     settings: Settings = get_settings()
 
-    def get_credentials(self) -> OAuth2LegacyCredentials:
+    def _get_app_access_token(self) -> dict:
+        """Acquire an access token for the Azure app"""
+        authority = f"https://login.microsoftonline.com/{self.settings.tenant_id}"
+        global_token_cache = _check_or_set_up_cache()
+        app = msal.ClientApplication(
+            self.settings.client_id,
+            authority=authority,
+            token_cache=global_token_cache,
+        )
+
+        accounts = app.get_accounts(username=self.settings.username)
+
+        if accounts:
+            logger.info("Account(s) exists in cache, probably with token too. Let's try.")
+            result = app.acquire_token_silent([self.settings.scope], account=accounts[0])
+        else:
+            logger.info("No suitable token exists in cache. Let's initiate interactive login.")
+            auth_code_flow = app.initiate_auth_code_flow(scopes=[self.settings.scope], login_hint=self.settings.username)
```

Comment:
Looks like you can set how long this token is valid for, might be worth seeing if we can make it a day so that you don't have to do it every hour
```suggestion
            auth_code_flow = app.initiate_auth_code_flow(scopes=[self.settings.scope], login_hint=self.settings.username, max_age=60*60*24)
```